### PR TITLE
Add extra validation checks for trace payloads

### PR DESCRIPTION
**Pull request by Cawllec**
## Goal

This further tightens the envelope on what a valid trace payload looks like, outside the scope of the very well defined payload schema.

This new mechanism can also be used to add pre-defined checks for other payload types.  We could add a superclass for the validator if necessary, allowing us to expand on it in the future if necessary, but that falls outside the scope of this PR.

## Tests

Unit tests have been added, as well as the existing trace tests being modified to reflect the new validation.

### CODE REVIEW

File: lib/features/steps/request_assertion_steps.rb [hunk 1]
```diff
@@ -56,6 +57,36 @@ def verify_schema_matches(list, list_name)
...
+      unless validator.success
+        $stdout.puts "\n"
+        $stdout.puts "\e[31m--- #{list_name} #{index} failed validation with the following errors:\e[0m"
+        validator.errors.each do |error|
+          $stdout.puts "\e[31m#{error}\e[0m"
+        end
+        $stdout.puts "\n"
+      end
+    end
+    raise Test::Unit::AssertionFailedError.new("A number of #{list_name} payloads failed validation.  A full list of the errors can be found above")
```
Comment (twometresteve) [inline lib/features/steps/request_assertion_steps.rb:86]:
```suggestion
    raise Test::Unit::AssertionFailedError.new("One or more #{list_name} payloads failed validation.  A full list of the errors can be found above")
```

**Review by twometresteve — APPROVED** [review]
Just a minor wording change suggestion.

File: lib/maze/schemas/trace_validator.rb [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        @success = true
+        regex_comparison('resourceSpans.0.scopeSpans.0.spans.0.spanId', '^[A-Fa-f0-9]{16}$')
+        regex_comparison('resourceSpans.0.scopeSpans.0.spans.0.traceId', '^[A-Fa-f0-9]{32}$')
+        regex_comparison('resourceSpans.0.scopeSpans.0.spans.0.kind', '^[A-Za-z_]+')
+        regex_comparison('resourceSpans.0.scopeSpans.0.spans.0.startTimeUnixNano', '^[0-9]+$')
+        regex_comparison('resourceSpans.0.scopeSpans.0.spans.0.endTimeUnixNano', '^[0-9]+$')
+        element_contains('resourceSpans.0.resource.attributes', 'os.type')
+        element_contains('resourceSpans.0.resource.attributes', 'os.name')
+        element_contains('resourceSpans.0.resource.attributes', 'deployment.environment', 'stringValue', ['development', 'production'])
+        element_contains('resourceSpans.0.resource.attributes', 'service.name')
```
Comment (gingerbenw) [inline lib/maze/schemas/trace_validator.rb:36]: this field is not applicable to browsers:

https://smartbear.atlassian.net/wiki/spaces/PROD/pages/3459842907/ROAD-1666+Browser+performance+-+Platforms+PD#Resource-attributes

File: lib/maze/schemas/trace_validator.rb [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        @success = nil
+        @errors = []
+      end
+
+      # Runs the validation against the trace given
+      def validate
+        @success = true
+        regex_comparison('resourceSpans.0.scopeSpans.0.spans.0.spanId', '^[A-Fa-f0-9]{16}$')
+        regex_comparison('resourceSpans.0.scopeSpans.0.spans.0.traceId', '^[A-Fa-f0-9]{32}$')
+        regex_comparison('resourceSpans.0.scopeSpans.0.spans.0.kind', '^[A-Za-z_]+')
```
Comment (imjoehaines) [inline lib/maze/schemas/trace_validator.rb:30]: `kind` is supposed to be an integer: 

> Values of enum fields MUST be encoded as integer values. Unlike the standard Protobuf JSON Mapping, which allows values of enum fields to be encoded as either integer values or as enum name strings, only integer enum values are allowed in OTLP JSON Protobuf Encoding; the enum name strings MUST NOT be used. For example, the kind field with a value of SPAN_KIND_SERVER in a Span can be represented like this: { “kind”: 2, ... }
>
> https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#json-protobuf-encoding

The mapping is:

```
Unspecified = 0
Internal = 1
Server = 2
Client = 3
Producer = 4
Consumer = 5
```

So a valid `kind` is an integer `>= 0 && <= 5`

File: lib/maze/schemas/trace_validator.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        regex_comparison('resourceSpans.0.scopeSpans.0.spans.0.startTimeUnixNano', '^[0-9]+$')
+        regex_comparison('resourceSpans.0.scopeSpans.0.spans.0.endTimeUnixNano', '^[0-9]+$')
+        element_contains('resourceSpans.0.resource.attributes', 'os.type')
+        element_contains('resourceSpans.0.resource.attributes', 'os.name')
```
Comment (gingerbenw) [inline lib/maze/schemas/trace_validator.rb:34]: neither `os.type` or `os.name` are valid for browsers either I'm afraid!

File: test/fixtures/payload-helpers/features/support/send_request.rb [hunk 5]
```diff
@@ -62,6 +62,163 @@ def send_request(request_type, mock_api_port = 9339)
...
+                    "stringValue":"com.bugsnag.Fixture"
+                  }
+                }
+              ]
+            }
+          }
+        ]
+      }
+    },
+    'mobile-trace' => {
```
Comment (gingerbenw) [inline test/fixtures/payload-helpers/features/support/send_request.rb:221]:
```suggestion
    'browser-trace' => {
```

File: test/fixtures/payload-helpers/features/traces_support.feature [hunk 6]
```diff
@@ -1,14 +1,21 @@
...
-        And the trace payload field "resourceSpans.0.resource.attributes.0.key" equals "telemetry.sdk.version"
-        And the trace payload field "resourceSpans.0.resource.attributes.0.value.stringValue" equals "0.0"
         And The HTTP response header "Bugsnag-Sampling-Probability" equals "1"
+        And the trace payload field "resourceSpans.0.resource.attributes.0.key" equals "device.id"
+        And the trace payload field "resourceSpans.0.resource.attributes.0.value.stringValue" equals "cd5c48566a5ba0b8597dca328c392e1a7f98ce86"
 
-        # Sampling probability can be set for all subsequent trace requests
+        # A browser trace
         Then I discard the oldest trace
+        And I send a "mobile-trace"-type request
```
Comment (gingerbenw) [inline test/fixtures/payload-helpers/features/traces_support.feature:13]:
```suggestion
        And I send a "browser-trace"-type request
```

**Review by gingerbenw — APPROVED** [review]
looks good, but would suggest updating the mobile-test name 👍🏻

**Pull request merged**